### ISSUE TITLE
Communication of vectors gives segfault in BOUT++ 4

### ISSUE THREAD

**Issue by loeiten** [boutproject/BOUT-dev#522]
Communication of `Vector3D` gives segfault in BOUT++-4

To illustrate the bug, condiser the following MWE

`testVec.hxx`

```cpp
#ifndef __VECTEST_H__
#define __VECTEST_H__

#include <bout/physicsmodel.hxx>
#include <vecops.hxx>         // Gives the vec diff options

class VecTest : public PhysicsModel {
protected:
  int init(bool restarting);
  int rhs(BoutReal t);

public:
  Field3D n;
  Vector3D gradPerpN;
  string ownOpType;
};

#endif
```

`testVec.cxx`

```cpp
#include "vecTest.hxx"

int VecTest::init(bool restarting) {
  TRACE("Halt in VecTest::init");
    SOLVE_FOR(n);
  return 0;
}

int VecTest::rhs(BoutReal t) {
  TRACE("Halt in VecTest::rhs");
  mesh->communicate(n);
  gradPerpN = Grad_perp(n);
  mesh->communicate(gradPerpN);
  ddt(n) = 0;
  return 0;
}

BOUTMAIN(VecTest);
```

`BOUT.inp`

```md
# Root options
###############################################################################
nout = 1         # Number of output timesteps
timestep = 1e-10 # Time between outputs

dump_format="nc" # Write NetCDF format files

# Setting the z coordinate
ZMIN = 0.0
ZMAX = 1.0 # dz = 2*pi(ZMAX - ZMIN)/(MZ - 1)

# Number of guard cells
MXG = 1
MYG = 1
###############################################################################


# Mesh options
###############################################################################
[mesh]
# Number of points
nx = 16
ny = 16
nz = 16
# Puts the boundaries half a step outside the last gridpoints
symmetricGlobalY=true
symmetricGlobalX=true
# Points 0 <= x < ixseps1 are inside separatrix, ixseps1 <= x < ngx outside
# --------Non-periodic----------
ixseps1 = -1
ixseps2 = -1
# ------------------------------
###############################################################################


# Methods options
###############################################################################
# Methods used for the radial (x) derivative terms
[ddx]
first  = C2 # d/dx (f)
second = C2 # d^2/dx^2 (f)
upwind = U1 # u*d/dx (f)

# Methods used for the axial (y) derivative terms
[ddy]
first  = C2 # d/dy (f)
second = C2 # d^2/dy^2 (f)
upwind = U1 # u*d/dy (f)

# Methods used for the azimuthal (z) derivative terms
[ddz]
first  = FFT # d/dz (f)
second = FTT # d^2/dz^2 (f)
###############################################################################


# Solver settings
###############################################################################
[solver]
# The solver advances the system in time
type = cvode             # Which solver to use (cvode should be same as pvode)
mms  = false             # false by default

atol = 1.0e-10 # absolute tolerance
rtol = 1.0e-5  # relative tolerance

# cvode, pvode and rk4 options
# NOTE: If this is set too high, the computer MIGHT interpret it as 0
mxstep = 100000000
###############################################################################


# Specifying the switches
###############################################################################
[output]
floats = false # floats = false => output in double
###############################################################################


# Settings for the variables
###############################################################################
[n]
scale    = 1.0
function = 0.0

# Boundaries
bndry_xin   = dirichlet_o4(0.0)
bndry_xout  = dirichlet_o4(0.0)
bndry_yup   = dirichlet_o4(0.0)
bndry_ydown = dirichlet_o4(0.0)
###############################################################################
```

executed with

```
make clean && make && mpirun -np 4 ./vecTest
```

yields the following segfault:

```
Error encountered
====== Exception thrown ======

****** SEGMENTATION FAULT CAUGHT ******

====== Back trace ======
 -> Mesh::communicate(FieldGroup&) on line 184 of 'mesh.cxx'
 -> Halt in VecTest::rhs on line 10 of 'vecTest.cxx'
====== Exception path ======
[bt] #1 ./vecTest() [0x45c8af]
BoutException::BoutException(char const*, ...) at /home/cooluser/BOUT-dev/src/sys/boutexception.cxx:102
[bt] #2 ./vecTest() [0x46800e]
bout_signal_handler(int) at /home/cooluser/BOUT-dev/src/bout++.cxx:501 (discriminator 2)
[bt] #3 /lib/x86_64-linux-gnu/libc.so.6(+0x354b0) [0x7f13b4d274b0]
?? ??:0
[bt] #4 ./vecTest() [0x48802c]
Array<double>::operator[](int) at /home/cooluser/BOUT-dev/src/mesh/impls/bout/../../../../include/bout/array.hxx:234 (discriminator 4)
[bt] #5 ./vecTest() [0x489a84]
BoutMesh::send(FieldGroup&) at /home/cooluser/BOUT-dev/src/mesh/impls/bout/boutmesh.cxx:954
[bt] #6 ./vecTest() [0x434e87]
Mesh::communicate(FieldGroup&) at /home/cooluser/BOUT-dev/src/mesh/mesh.cxx:190
[bt] #7 ./vecTest() [0x406d06]
void Mesh::communicate<Vector3D>(Vector3D&) at /home/cooluser/BOUT-dev/include/bout/mesh.hxx:193
[bt] #8 ./vecTest() [0x4523f0]
PhysicsModel::runRHS(double) at /home/cooluser/BOUT-dev/src/physics/physicsmodel.cxx:38
[bt] #9 ./vecTest() [0x459a48]
Solver::run_rhs(double) at /home/cooluser/BOUT-dev/src/solver/solver.cxx:1299
[bt] #10 ./vecTest() [0x45a409]
Solver::solve(int, double) at /home/cooluser/BOUT-dev/src/solver/solver.cxx:523
[bt] #11 ./vecTest() [0x406eec]
main at /home/cooluser/Documents/Utdanning/Danmarks_Tekniske_Universitet/Ph.D./CELMA/celmaWithBoussinesqApprox/delme5/vecTest.cxx:18 (discriminator 10) 
[bt] #12 /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7f13b4d12830]
?? ??:0
[bt] #13 ./vecTest() [0x4069a9]
_start at ??:?
```

**Comment by bendudson** [comment]
Sorry. I think this is fixed in PR #511

**Comment by bendudson** [comment]
Already fixed in next (PR #491, not #511), not yet fixed in master.

**Comment by loeiten** [comment]
Ah, already fixed. Nothing is better than that.
Should commit 07d81d17c10455b25441ba321ade8e65ffc4c235 go to master as well as this is a bugfix?
